core/src/main/java/com/google/cloud/sql/core/CloudSqlConnectorRegistry.java

```diff
@@ -0,0 +1,361 @@
+    CloudSqlConnectorRegistry instance = getInstance();
+    return instance.getConnectionInfoCache(config).getSslData(instance.refreshTimeoutMs);
+  }
+
+  /** Returns preferred ip address that can be used to establish Cloud SQL connection. */
+  public static String getHostIp(ConnectionConfig config) throws IOException {
+    CloudSqlConnectorRegistry instance = getInstance();
+    return instance
+        .getConnectionInfoCache(config)
+        .getPreferredIp(config.getIpTypes(), instance.refreshTimeoutMs);
+  }
+
+  private static KeyPair generateRsaKeyPair() {
+    KeyPairGenerator generator;
+    try {
+      generator = KeyPairGenerator.getInstance("RSA");
+    } catch (NoSuchAlgorithmException err) {
+      throw new RuntimeException(
+          "Unable to initialize Cloud SQL socket factory because no RSA implementation is "
+              + "available.");
+    }
+    generator.initialize(RSA_KEY_SIZE);
+    return generator.generateKeyPair();
+  }
+
+  private static String getVersion() {
+    try {
+      Properties packageInfo = new Properties();
+      packageInfo.load(
+          CloudSqlConnectorRegistry.class
+              .getClassLoader()
+              .getResourceAsStream("com.google.cloud.sql/project.properties"));
+      return packageInfo.getProperty("version", "unknown");
+    } catch (IOException e) {
+      return "unknown";
+    }
+  }
+
+  /**
+   * Internal use only: Sets the default string which is appended to the SQLAdmin API client
+   * User-Agent header.
+   *
+   * <p>This is used by the specific database connector socket factory implementations to append
+   * their database name to the user agent.
+   */
+  public static void addArtifactId(String artifactId) {
+    String userAgent = artifactId + "/" + version;
+    if (!userAgents.contains(userAgent)) {
+      userAgents.add(userAgent);
+    }
+  }
+
+  /** Resets the values of User Agent fields for unit tests. */
+  @VisibleForTesting
+  static void resetUserAgent() {
+    cloudSqlConnectorRegistry = null;
+    userAgents.clear();
+    setApplicationName("");
+  }
+
+  /** Returns the default string which is appended to the SQLAdmin API client User-Agent header. */
+  static String getUserAgents() {
+    String ua = String.join(" ", userAgents);
+    String appName = getApplicationName();
+    if (!Strings.isNullOrEmpty(appName)) {
+      ua = ua + " " + appName;
+    }
+    return ua;
+  }
+
+  /** Returns the current User-Agent header set for the underlying SQLAdmin API client. */
+  private static String getApplicationName() {
+    return System.getProperty(USER_TOKEN_PROPERTY_NAME, "");
+  }
+
+  /**
+   * Adds an external application name to the user agent string for tracking. This is known to be
+   * used by the spring-cloud-gcp project.
+   *
+   * @throws IllegalStateException if the SQLAdmin client has already been initialized
+   */
+  public static void setApplicationName(String applicationName) {
```

Comment:
What shall we do about this? I forget what the plan here was.

Reply:
We're going to deprecate this method in favor of `ConnectorRegistry.addApplicationId(String id)`.